### PR TITLE
Sort: Write Reference Tables where possible

### LINKED ISSUE #2102: Huge memory usage increase?
I noticed that the memory consumption of Hyrise’s TPC-H benchmark may have increased drastically. When executing
`./hyriseBenchmarkTPCH -q 2 -s 0.1 `
from commit 8851c50467ad8d1bc792434fc55b2c0066dd48a6 (Thu Jan 23 14:12:15 2020 +010) `htop` shows ~3.4 GB of virtual memory and a resident set size of ~580 MB.

When executing the benchmark with the current master the virtual memory increases to ~24.9 GB and the resident set size increases to ~2 GB. Jemalloc reports stats.allocated = 21'800'995'904 after having executed query 2 of the TPC-H benchmark.

### PR DESCRIPTION
**Pull request by mrks**
In most cases, it is unnecessary to materialize the result of a sort. This adds a method that returns reference tables instead of data tables.

Exception: If a column in the input references different tables or columns (e.g., as the result of a union), we still have to materialize. Also, we allow the caller to request materialization (e.g., when creating sorted tables for the benchmarks). For this, this PR fixes the materialization (fixes #2102).

The Sort test (which previously operated on tables containing only three lines) was entirely rewritten.

Microbenchmarks - note that the fixed materialization makes even the non-reference version in this branch almost as fast as the reference version:
```
Comparing /home/Markus.Dreseler/hyrise/build-release/hyriseMicroBenchmarks to ./hyriseMicroBenchmarks
Benchmark                                                        Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------
BM_Sort/100                                                   -0.4003         -0.4003          7934          4758          7934          4758
BM_Sort/10000                                                 -0.1438         -0.1438       1464272       1253674       1464295       1253687
BM_Sort/1000000                                               -0.2147         -0.2147     227946520     179003477     227951168     179004475
BM_SortTwoColumns/100                                         -0.4160         -0.4161          8056          4704          8055          4704
BM_SortTwoColumns/10000                                       -0.1614         -0.1614       1501766       1259356       1501780       1259364
BM_SortTwoColumns/1000000                                     -0.1624         -0.1624     222237984     186138550     222239366     186142513
BM_SortWithNullValues/100                                     -0.4066         -0.4066          8174          4851          8174          4851
BM_SortWithNullValues/10000                                   -0.1617         -0.1617       1269751       1064426       1269773       1064437
BM_SortWithNullValues/1000000                                 -0.1882         -0.1881     190632820     154764175     190634397     154767434
BM_SortWithReferenceSegments/100                              -0.3330         -0.3329          8284          5525          8283          5525
BM_SortWithReferenceSegments/10000                            -0.2602         -0.2601       1053693        779524       1053614        779539
BM_SortWithReferenceSegments/1000000                          -0.4424         -0.4422     214130402     119409132     214094124     119411420
BM_SortWithReferenceSegmentsTwoColumns/100                    -0.2272         -0.2272         10933          8449         10933          8449
BM_SortWithReferenceSegmentsTwoColumns/10000                  -0.1593         -0.1593       1783649       1499560       1783647       1499433
BM_SortWithReferenceSegmentsTwoColumns/1000000                -0.2727         -0.2727     357221723     259798845     357226980     259801061
BM_SortWithStrings/100                                        -0.1311         -0.1311         46536         40435         46536         40435
BM_SortWithStrings/10000                                      -0.0690         -0.0690       7975963       7425759       7976112       7425689
BM_SortWithStrings/1000000                                    -0.1416         -0.1416    1453135490    1247404099    1453144154    1247347292
```

TPC-H:
```diff
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | Benchmark      | prev. iter/s       | runs  | new iter/s         | runs  | change [%] | p-value |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | TPC-H 01       | 1.2703073024749756 | 77    | 1.286609411239624  | 78    | +1%        |  0.0033 |
+| TPC-H 02       | 78.9946517944336   | 4741  | 91.70992279052734  | 5503  | +16%       |  0.0000 |
 | TPC-H 03       | 9.271662712097168  | 557   | 9.2130126953125    | 553   | -1%        |  0.0056 |
 | TPC-H 04       | 8.405821800231934  | 505   | 8.14923095703125   | 489   | -3%        |  0.0000 |
-| TPC-H 05       | 4.2524518966674805 | 256   | 3.926089286804199  | 236   | -8%        |  0.0000 |
 | TPC-H 06       | 270.4285888671875  | 16226 | 278.2364807128906  | 16695 | +3%        |  0.0000 |
 | TPC-H 07       | 9.815805435180664  | 590   | 9.74395751953125   | 585   | -1%        |  0.3007 |
 | TPC-H 08       | 7.586434364318848  | 456   | 7.447839736938477  | 447   | -2%        |  0.0000 |
 | TPC-H 09       | 1.7536722421646118 | 106   | 1.745798945426941  | 105   | -0%        |  0.1745 |
+| TPC-H 10       | 3.3918466567993164 | 204   | 3.612431049346924  | 217   | +7%        |  0.0000 |
 | TPC-H 11       | 40.326927185058594 | 2420  | 38.81638717651367  | 2330  | -4%        |  0.0000 |
 | TPC-H 12       | 17.037710189819336 | 1023  | 17.01153564453125  | 1021  | -0%        |  0.2073 |
 | TPC-H 13       | 2.5991594791412354 | 156   | 2.656994342803955  | 160   | +2%        |  0.0000 |
 | TPC-H 14       | 43.671085357666016 | 2621  | 45.43584442138672  | 2727  | +4%        |  0.0000 |
 | TPC-H 15       | 121.0084228515625  | 7261  | 124.7322998046875  | 7484  | +3%        |  0.0000 |
+| TPC-H 16       | 9.685030937194824  | 582   | 10.211758613586426 | 613   | +5%        |  0.0000 |
 | TPC-H 17       | 5.571112155914307  | 335   | 5.3744964599609375 | 323   | -4%        |  0.0000 |
 | TPC-H 18       | 1.3789422512054443 | 83    | 1.3681787252426147 | 83    | -1%        |  0.3782 |
 | TPC-H 19       | 10.215389251708984 | 613   | 9.940498352050781  | 597   | -3%        |  0.0000 |
 | TPC-H 20       | 32.61243438720703  | 1957  | 31.788984298706055 | 1908  | -3%        |  0.0000 |
 | TPC-H 21       | 1.4128497838974    | 85    | 1.386466145515442  | 84    | -2%        |  0.0001 |
 | TPC-H 22       | 18.334503173828125 | 1101  | 19.102487564086914 | 1147  | +4%        |  0.0000 |
 | geometric mean |                    |       |                    |       | +1%        |         |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
```

### CODE REVIEW

**Comment by Bouncner** [comment]
Looking at the PQP of Q5, that should be an outlier, right?
The other gains make sense to me.

**Comment by mrks** [comment]
I think so, too.

File: src/lib/operators/sort.cpp [hunk 1]
```diff
@@ -8,8 +8,8 @@ using namespace opossum;  // NOLINT
 
 // Given an unsorted_table and a pos_list that defines the output order, this materializes all columns in the table,
 // creating chunks of output_chunk_size rows at maximum.
-std::shared_ptr<Table> materialize_output_table(const std::shared_ptr<const Table>& unsorted_table,
-                                                const RowIDPosList& pos_list, const ChunkOffset output_chunk_size) {
+std::shared_ptr<Table> write_materialized_output_table(const std::shared_ptr<const Table>& unsorted_table,
+                                                       RowIDPosList pos_list, const ChunkOffset output_chunk_size) {
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/sort.cpp:12]: Yes.

File: resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl [hunk 2]
```diff
@@ -1,3 +1,3 @@
 table_name|chunk_id|column_id|order_mode
 string|int|int|string
-int_int|0|1|AscendingNullsFirst
+int_int|0|1|Ascending
```
[the remark this answers is not in the document]
Comment (mrks) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]:
```
              Whether a sort key value that is null is considered greater
              or less than a non-null value is implementation-defined, but
              all sort key values that are null shall either be considered
              greater than all non-null values or be considered less than
              all non-null values.
```

The whole nulls first / last thing is old. Like... really old: https://github.com/hyrise/hyrise/pull/250#discussion_r139668590

It seems to be in there just to support `ORDER BY foo DESC NULLS LAST`, which I have never used. Remove it?

[the remark this answers is not in the document]
Comment (Bouncner) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]: Did not know that SQL supports such things. When it's SQL, I wouldn't remove it.

[the remark this answers is not in the document]
Comment (mrks) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]: It's non-standard.

[the remark this answers is not in the document]
Comment (Bouncner) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]: Then I am in favor of dropping it.

[the remark this answers is not in the document]
Comment (mrks) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]: Moved into #2095

**Comment by julianmenzler** [comment]
Still on it. Will finish review today.

File: src/test/operators/sort_test.cpp [hunk 3]
```diff
@@ -1,259 +1,188 @@
...
-  EXPECT_TABLE_EQ_ORDERED(sort->get_output(), expected_result);
+  EXPECT_EQ(sort.get_output()->type(), TableType::References);
 }
 
-TEST_P(OperatorsSortTest, DescendingSortOfOneColumnWithNullsLast) {
-  std::shared_ptr<Table> expected_result =
-      load_table("resources/test_data/tbl/int_float_null_sorted_desc_nulls_last.tbl", 2);
+TEST_F(SortTest, InputReferencesDifferentTables) {
+  // When a single column in a table references different tables, we cannot output sorted ReferenceSegments.
+  // This tests simulates the output of a union on the first column.
```
Comment (julianmenzler) [inline src/test/operators/sort_test.cpp:132]:
```suggestion
  // This test simulates the output of a union on the first column.
```

File: src/test/operators/sort_test.cpp [hunk 4]
```diff
@@ -1,259 +1,188 @@
...
-      _table_wrapper_null,
-      std::vector<SortColumnDefinition>{SortColumnDefinition{ColumnID{0}, OrderByMode::DescendingNullsLast}}, 2u);
-  sort->execute();
+  const auto second_table = load_table("resources/test_data/tbl/sort/a_asc.tbl", 10);
+  const auto second_table_wrapper = std::make_shared<TableWrapper>(second_table);
+  second_table_wrapper->execute();
 
-  EXPECT_TABLE_EQ_ORDERED(sort->get_output(), expected_result);
-}
+  const auto reference_table = std::make_shared<Table>(
```
Comment (julianmenzler) [inline src/test/operators/sort_test.cpp:138]: Suggestion: Name it `union_table`

File: src/lib/operators/sort.cpp [hunk 5]
```diff
@@ -147,22 +256,54 @@ std::shared_ptr<const Table> Sort::_on_execute() {
...
       previously_sorted_pos_list = sort_impl.sort(previously_sorted_pos_list);
+    });
+  }
 
-      if (is_last_sorting_step) {
-        // This is inside the for loop so that we do not have to resolve the type again
-        sorted_table = materialize_output_table(input_table, *previously_sorted_pos_list, _output_chunk_size);
+  // We have to materialize the output (i.e., write ValueSegments) if (a) it is requested by the user or (b) a column
+  // in the table references multiple tables itself (see write_reference_output_table for details). The latter can only
+  // occur if there is more than one ReferenceSegment in an input chunk.
```
[the remark this answers is not in the document]
Comment (julianmenzler) [reply src/lib/operators/sort.cpp:264]: Consider adding:
> (c) a column in the table references multiple columns of a single table.

File: src/lib/operators/sort.cpp [hunk 6]
```diff
@@ -94,15 +98,114 @@ std::shared_ptr<Table> materialize_output_table(const std::shared_ptr<const Tabl
...
+  // Ceiling of integer division
+  const auto div_ceil = [](auto x, auto y) { return (x + y - 1u) / y; };
+  const auto output_chunk_count = div_ceil(input_pos_list.size(), output_chunk_size);
+  Assert(input_pos_list.size() == unsorted_table->row_count(), "Mismatching size of input table and PosList");
+
+  // Vector of segments for each chunk
+  auto output_segments_by_chunk = std::vector<Segments>(output_chunk_count, Segments(column_count));
+
+  if (input_pos_list.size() <= output_chunk_size && !resolve_indirection) {
+    // Shortcut: No need to copy RowIDs if input_pos_list is small enough and we do not need to resolve the indirection
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/sort.cpp:127]: Moved the predicate. I don't agree with the suggestion about the comment. If we explain the condition, we should explain it fully and not expect the reader to realize that the other half is in the code.

File: src/lib/operators/sort.cpp [hunk 7]
```diff
@@ -94,15 +98,114 @@ std::shared_ptr<Table> materialize_output_table(const std::shared_ptr<const Tabl
...
+    auto& output_segments = output_segments_by_chunk.at(0);
+    for (auto column_id = ColumnID{0u}; column_id < column_count; ++column_id) {
+      output_segments[column_id] = std::make_shared<ReferenceSegment>(unsorted_table, column_id, output_pos_list);
+    }
+  } else {
+    for (ColumnID column_id{0u}; column_id < column_count; ++column_id) {
+      // We write the output ReferenceSegments column by column. This means that even if input ReferenceSegments share a
+      // PosList, the output will contain independent PosLists. While this is slightly more expensive to generate and
+      // slightly less efficient for following operators, we assume that the lion's share of the work has been done
+      // before the sort operator is executed and that the relative cost of this is acceptable. In the future, this
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/sort.cpp:138]: Started with "To keep the implementation simple"

File: src/lib/operators/sort.cpp [hunk 8]
```diff
@@ -94,15 +98,114 @@ std::shared_ptr<Table> materialize_output_table(const std::shared_ptr<const Tabl
...
+        DebugAssert(!output_pos_list->empty(), "Asked to write empty output_pos_list");
+        output_segments_by_chunk.at(output_chunk_id)[column_id] =
+            std::make_shared<ReferenceSegment>(referenced_table, referenced_column_id, output_pos_list);
+        ++output_chunk_id;
+
+        output_pos_list = std::make_shared<RowIDPosList>();
+        if (output_chunk_id < output_chunk_count) {
+          output_pos_list->reserve(output_chunk_size);
+        }
+      };
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/sort.cpp:165]: added

> write_output_pos_list creates an output reference segment for a given ChunkID, ColumnID and PosList

File: src/lib/operators/sort.cpp [hunk 9]
```diff
@@ -94,15 +98,114 @@ std::shared_ptr<Table> materialize_output_table(const std::shared_ptr<const Tabl
...
+
+      for (auto input_pos_list_offset = size_t{0}; input_pos_list_offset < input_pos_list.size();
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/sort.cpp:167]: > Iterate over rows in sorted input pos list, dereference them if necessary, and write a chunk every `output_chunk_size` rows

File: src/test/operators/sort_test.cpp [hunk 10]
```diff
@@ -1,259 +1,188 @@
...
+                           // Output chunk size
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, false, 40,                  Sort::ForceMaterialization::No,  "a_asc_b_desc.tbl"},      // NOLINT
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, true,  40,                  Sort::ForceMaterialization::No,  "a_asc_b_desc.tbl"},      // NOLINT
 
-  EXPECT_TABLE_EQ_ORDERED(sort->get_output(), expected_result);
-}
-
-TEST_P(OperatorsSortTest, DescendingSortOfOneColumnWithNull) {
-  std::shared_ptr<Table> expected_result = load_table("resources/test_data/tbl/int_float_null_sorted_desc.tbl", 2);
+                           // Empty input tables
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/operators/sort_test.cpp:102]: done

File: src/test/operators/sort_test.cpp [hunk 11]
```diff
@@ -1,259 +1,188 @@
...
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}},                                                                       true,  true,  Chunk::DEFAULT_SIZE, Sort::ForceMaterialization::No,  "empty.tbl"},             // NOLINT
 
-  auto sort = std::make_shared<Sort>(
-      _table_wrapper_null,
-      std::vector<SortColumnDefinition>{SortColumnDefinition{ColumnID{0}, OrderByMode::Descending}}, 2u);
-  sort->execute();
+                           // Forced materialization
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, false, Chunk::DEFAULT_SIZE, Sort::ForceMaterialization::Yes, "a_asc_b_desc.tbl"},      // NOLINT
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, false, 33,                  Sort::ForceMaterialization::Yes, "a_asc_b_desc.tbl"},      // NOLINT
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/operators/sort_test.cpp:108]: Mmh, it's difficult to order things with multiple dimensions. We already tested different output chunk sizes above. This block checks whether things still work when forced materialization enters the stage. I'd leave it the way it is.

File: src/test/operators/sort_test.cpp [hunk 12]
```diff
@@ -1,259 +1,188 @@
...
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, true,  Chunk::DEFAULT_SIZE, Sort::ForceMaterialization::Yes, "a_asc_b_desc.tbl"},      // NOLINT
+                           SortTestParam{{SortColumnDefinition{ColumnID{0}, OrderByMode::Ascending}, SortColumnDefinition{ColumnID{1}, OrderByMode::Descending}},           false, true,  33,                  Sort::ForceMaterialization::Yes, "a_asc_b_desc.tbl"}       // NOLINT
```
[the remark this answers is not in the document]
Comment (mrks) [reply src/test/operators/sort_test.cpp:110]: dito

File: src/lib/operators/sort.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (julianmenzler) [reply src/lib/operators/sort.cpp]: Alright. I'm also fine with that.

File: src/test/operators/sort_test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (julianmenzler) [reply src/test/operators/sort_test.cpp:112]: Okay, sounds rational.

**Comment by mrks** [comment]
Re-running CI right now. That doesn't look like an issue in this code. If this is ccache again, I am killing it for good.

File: src/test/operators/sort_test.cpp [hunk 13]
```diff
@@ -162,7 +170,8 @@ TEST_F(SortTest, InputReferencesDifferentColumns) {
   // Similarly to InputReferencesDifferentTables, we cannot build a ReferenceSegment that references different columns
   // in the same table.
 
-  const auto reference_table = std::make_shared<Table>(
+  // This is not just a normal union_table but something weird that you probably won't see in the wild
+  const auto weird_table = std::make_shared<Table>(
```
Comment (julianmenzler) [inline src/test/operators/sort_test.cpp:174]: Had to laugh on this one. Great! 😄

File: resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bensk1) [reply resources/test_data/tbl/meta_tables/meta_chunk_sort_orders_updated.tbl:3]: I was looking for arguments to keep the support for that and did not find any. But I found something else which is useless but interesting: https://docs.mendix.com/refguide/null-ordering-behavior

**Pull request merged**